h/api/models/document.py

```diff
@@ -182,6 +186,153 @@ def __repr__(self):
+    elif not docuri.document == document:
+        log.warn('Found DocumentURI with id %d does not match expected '
+                 'document with id %d', docuri.document_id, document.id)
+
+    docuri.updated = updated
+
+
+def create_or_update_document_meta(db,
+                                   claimant,
+                                   claimant_normalized,
+                                   type,
+                                   value,
+                                   document,
+                                   created,
+                                   updated):
+    """
+    Create or update a DocumentMeta with the given parameters.
+
+    If an equivalent DocumentMeta already exists in the database then its value
+    and updated time will be updated.
+
+    If no equivalent DocumentMeta exists in the database then a new one will be
+    created and added to the database.
+
+    To be considered "equivalent" an existing DocumentMeta must have the given
+    claimant and type, but its value, document and created and updated times
+    needn't match the given ones.
+
+    :param db: the database session
+    :type db: sqlalchemy.orm.session.Session
+
+    :param claimant: the value to use for the DocumentMeta's claimant attribute
+        if a new DocumentMeta is created
+    :type claimant: unicode
+
+    :param claimant_normalized: the value of the new or existing DocumentMeta's
+        claimant_normalized attribute
+    :type claimant_normalized: unicode
+
+    :param type: the value of the new or existing DocumentMeta's type attribute
+    :type type: unicode
+
+    :param value: the value to set the new or existing DocumentMeta's value
+        attribute to
+    :type value: unicode
+
+    :param document: the value to use for the DocumentMeta's document if a new
+        DocumentMeta is created
+    :type document: h.api.models.Document
+
+    :param created: the value to use for the DocumentMeta's created attribute
+        if a new DocumentMeta is created
+    :type created: datetime.datetime
+
+    :param updated: the value to set the new or existing DocumentMeta's updated
+        attribute to
+    :type updated: datetime.datetime
+
+    """
+    existing_dm = DocumentMeta.query.filter(
+        DocumentMeta.claimant_normalized == claimant_normalized,
+        DocumentMeta.type == type).one_or_none()
+
+    if existing_dm is None:
+        db.add(DocumentMeta(
+            _claimant=claimant,
+            _claimant_normalized=claimant_normalized,
+            type=type,
+            value=value,
+            document=document,
+            created=created,
+            updated=updated,
+        ))
+    else:
+        existing_dm.value = value
+        existing_dm.updated = updated
+        if not existing_dm.document == document:
+            log.warn('Found DocumentMeta with id %d does not match expected '
+                     'document with id %d', existing_dm.id,
+                     document.id)
```

Comment:
Adding the `existing_dm.document_id` might be good as well, as this `DocumentMeta` instance can move to another document in the future, so the log message might be more helpful when we include all three ids.